File: runme/mcgen/local/datan/solsrc/s4lsq.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

extern double CBFUNC gsfnct(double const P_T *x, integer const P_T *nr,
                            double const P_T *t);
extern double CBFUNC bwfnct(double const P_T *x, integer const P_T *nr,
                            double const P_T *t);

/* Table of constant values */

static integer c__1 = 1, c__19 = 19, c__21 = 21;
static double c_dp5 = .5;

main(void)
{
    /* Local variables */
    static integer i, ipl, npl, lcapt, list[2], ltx, lty, nws, nstep, iswit,
                  nr = 2;
    static double a[42] /* was [21][2] */, cx[4] /* was [2][2] */,
                  datcov[21], deltay[21], p, r, t[21], x[2], y[21], sigma,
                  scrat[4] /* was [2][2] */, datsx[21], xpl[100], ypl[100];
    static char capt[75], tx[75], ty[75];
    /* System generated locals */
    integer i__1;
/* identify program to user */
    printf("%s\n", "Program S4LSQ generates data");
    printf("%s\n", "corresponding to a Breit-Wigner law");
    printf("%s\n\n", "and fits a Breit-Wigner OR a Gaussian.");
/* ask for input */
    printf("%s\n", "Enter size of measurement errors (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - Fit to Breit-Wigner function");
    printf("%s\n", "2 - Fit to Gaussian");
    printf("%s", "> ");
    scanf("%li", &iswit);
/* generate data points corresponding to Breit-Wigner */
    x[0] = 0.;
    x[1] = 1.;
    rnstnr_(deltay, &c__21);
    for (i = 1; i <= 21; ++i) {
	t[i - 1] = (double) (i - 1) * .3 - 3.;
	y[i - 1] = bwfnct(x, &nr, &t[i - 1]);
	y[i - 1] += deltay[i - 1] * sigma;
	deltay[i - 1] = sigma;
    }
/* set first approximation of unknowns */
    x[0] = .5;
    x[1] = .5;
/* perform fit */
    nstep = 100;
    if (iswit == 1) {
	lsqnon_(bwfnct, t, y, deltay, &c__21, &nr, &nr, list, x, 
	       cx, &r, a, scrat, &nstep);
    } else {
	lsqnon_(gsfnct, t, y, deltay, &c__21, &nr, &nr, list, x, 
	       cx, &r, a, scrat, &nstep);
    }
    p = 1. - scchi2_(&r, &c__19);
    if (nstep < 0) {
        printf("%s%5li\n", "LSQNON ended with NSTEP = ", nstep);
    } else {
/* prepare graphics */
        memset(tx, ' ', 75);
        memset(ty, ' ', 75);
	strncpy(tx, "t", 1);
	strncpy(ty, "y", 1);
	ltx = 1;
	lty = 1;
	sprintf(capt, "x_1#=%6.2lf, x_2#=%6.2lf, M=%6.2lf, P=%8.6lf",
                x[0], x[1], r, p);
	if (iswit == 1) {
            strncpy(capt + 46, " (fit to Breit-Wigner) ", 23);
	} else {
            strncpy(capt + 46, " (fit to Gaussian)     ", 23);
	}
        memset(capt + 69, ' ', 6);
	lcapt = 69;
/* curve corresponding to solution */
	npl = 100;
	i__1 = npl;
	for (ipl = 1; ipl <= i__1; ++ipl) {
	    xpl[ipl - 1] = (double) ipl * .08 - 4.;
	    if (iswit == 1) {
		ypl[ipl - 1] = bwfnct(x, &nr, &xpl[ipl - 1]);
	    } else {
		ypl[ipl - 1] = gsfnct(x, &nr, &xpl[ipl - 1]);
	    }
	}
/* prepare data points for graphical presentation */
	for (i = 1; i <= 21; ++i) {
	    datsx[i - 1] = 0.;
	    datcov[i - 1] = 0.;
	}
/* ask for number of workstation */
        grnbws_();
        printf("%s\n", "Please, enter number of workstation:");
        printf("%s", "> ");
        scanf("%li", &nws);
        (void)getchar();
/* graphical output */
	grdtcv_(xpl, ypl, &npl, &c__1, &c_dp5, t, y, datsx, deltay, datcov,
               &c__21, tx, &ltx, ty, &lty, capt, &lcapt, &nws);
    }
    return 0;
} /* main */

/* ------------------------------------------------- */
double CBFUNC bwfnct(double const P_T *x, integer const P_T *nr,
                     double const P_T *t)
{
    /* System generated locals */
    double ret_val, d__1, d__2, d__3;

    /* Parameter adjustments */
    --x;

    /* Function Body */
/* Computing 2nd power */
    d__1 = x[2];
/* Computing 2nd power */
    d__2 = *t - x[1];
/* Computing 2nd power */
    d__3 = x[2];
    ret_val = d__1 * d__1 * 2. / (x[2] * 3.14159f * (d__2 * d__2 * 4. + d__3 *
	     d__3));
    return ret_val;
} /* bwfnct */

/* ------------------------------------------------- */
double CBFUNC gsfnct(double const P_T *x, integer const P_T *nr,
                     double const P_T *t)
{
    /* System generated locals */
    double ret_val, d__1;

    /* Parameter adjustments */
    --x;

    /* Function Body */
/* Computing 2nd power */
    d__1 = (*t - x[1]) / x[2];
    ret_val = .39894228 / x[2] * exp(d__1 * d__1 * -.5);
    return ret_val;
} /* gsfnct */

File: runme/mcgen/local/datan/solsrc/s4lsq.f
      PROGRAM S4LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXNR=2,MAXN=21,N=21,T0=-3.D0,DELTAT=0.3D0)
      DIMENSION T(MAXN),DELTAY(MAXN),CX(MAXNR,MAXNR),A(MAXN,MAXNR)
      DIMENSION Y(MAXN),X(MAXNR),SCRAT(MAXNR,MAXNR)
      CHARACTER*75 TX,TY,CAPT
      DIMENSION DATSX(MAXN),DATCOV(MAXN)
      DIMENSION XPL(100),YPL(100)
      DIMENSION LIST(MAXNR)
      EXTERNAL BWFNCT,GSFNCT
      DATA NR /2/
C identify program to user
      WRITE(*,*)' Program S4LSQ generates data'
      WRITE(*,*)' corresponding to a Breit-Wigner law'
      WRITE(*,*)' and fits a Breit-Wigner  OR a Gaussian'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter size of measurement errors (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - Fit to Breit-Wigner function'
      WRITE(*,*)' 2 - Fit to Gaussian'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
C generate data points corresponding to Breit-Wigner
      X(1)=0.0D0
      X(2)=1.0D0
      CALL RNSTNR(DELTAY,N)
      DO 10 I=1,N
        T(I)=T0+DBLE(I-1)*DELTAT
        Y(I)=BWFNCT(X,NR,T(I))
        Y(I)=Y(I)+DELTAY(I)*SIGMA
        DELTAY(I)=SIGMA
10    CONTINUE
C set first approximation of unknowns
      X(1)=0.5D0
      X(2)=0.5D0
C perform fit
      NSTEP=100
      IF(ISWIT.EQ.1) THEN
        CALL LSQNON(BWFNCT,T,Y,DELTAY,N,NR,NR,LIST,X,CX,R,A,SCRAT,
     +  NSTEP)
      ELSE
        CALL LSQNON(GSFNCT,T,Y,DELTAY,N,NR,NR,LIST,X,CX,R,A,SCRAT,
     +  NSTEP)
      END IF
      P=1.D0-SCCHI2(R,N-NR)
      IF(NSTEP.LT.0) THEN
        WRITE(*,'(A,I5)')' LSQNON ended with NSTEP = ',NSTEP
      ELSE
C prepare graphics
        TX='t'
        TY='y'
        LTX=1
        LTY=1
        CAPT='x_1#=******, x_2#=******, M=******, P=********'
        WRITE(CAPT(6:11),'(F6.2)')X(1)
        WRITE(CAPT(19:24),'(F6.2)')X(2)
        WRITE(CAPT(29:34),'(F6.2)')R
        WRITE(CAPT(39:46),'(F8.6)')P
        IF(ISWIT.EQ.1) THEN
          WRITE(CAPT(47:69),'(A)')' (fit to Breit-Wigner)'
        ELSE
          WRITE(CAPT(47:69),'(A)')' (fit to Gaussian)    '
        END IF
        LCAPT=69
C curve corresponding to solution
        NPL=100
        DO 20 IPL=1,NPL
          XPL(IPL)=-4.D0+0.08D0*DBLE(IPL)
          IF(ISWIT.EQ.1) THEN
            YPL(IPL)=BWFNCT(X,NR,XPL(IPL))
          ELSE
            YPL(IPL)=GSFNCT(X,NR,XPL(IPL))
          END IF
20      CONTINUE
C prepare data points for graphical presentation
        DO 30 I=1,N
          DATSX(I)=0.D0
          DATCOV(I)=0.D0
30      CONTINUE
C ask for number of workstation
        CALL GRNBWS()
        WRITE(*,*)' Please, enter number of workstation:'
        WRITE(*,*)' >'
        READ(*,*)NWS
C graphical output
        CALL GRDTCV(XPL,YPL,NPL,1,.5D0,T,Y,DATSX,DELTAY,DATCOV,N,
     +  TX,LTX,TY,LTY,CAPT,LCAPT,NWS)
      END IF
      END
C-------------------------------------------------
      DOUBLE PRECISION FUNCTION BWFNCT(X,NR,T)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION X(*)
      BWFNCT=2.D0*X(2)**2/((3.14159*X(2))*(4.D0*(T-X(1))**2+X(2)**2))
      END
C-------------------------------------------------
      DOUBLE PRECISION FUNCTION GSFNCT(X,NR,T)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION X(*)
      GSFNCT=(.39894228D0/X(2))*EXP(-0.5D0*((T-X(1))/X(2))**2)
      END
